src/librustc_resolve/lib.rs

```diff
@@ -5030,11 +5030,13 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
             if let (
                 Ok(snippet),
                 NameBindingKind::Import { directive, ..},
-                _dummy @ false,
+                false,
+                false,
             ) = (
                 cm.span_to_snippet(binding.span),
                 binding.kind.clone(),
                 binding.span.is_dummy(),
+                binding.span.ctxt().outer().expn_info().is_some(),
```

Comment:
If you rely on `source.span.hi().0 - binding.span.lo().0` being a valid index in `binding.span`, then you should check at least the next conditions - `!source.span.is_dummy() && !binding.span.is_dummy() && !from_macro(source.span) && !from_macro(binding.span)`.

Reply:
Actually, directly checking that `source.span.hi().0 - binding.span.lo().0` doesn't underflow and results in a valid index into `cm` would be the most reliable alternative.
This way we certainly don't get an ICE, but also keep the diagnostics for "good" macros.